index.js

```diff
@@ -142,59 +154,177 @@ function updateSchemesAndHost(req) {
+/**
+ * Apply this **last**!
+ *
+ * (as the very last middleware of your express app)
+ *
+ * @description `request` middleware.
+ * Applies to the `app` you provided in `injectResponseMiddleware`
+ *
+ * @returns void
+ */
+function injectRequestMiddleware() {
+  /** middleware to handle REQUESTS */
+  // eslint-disable-next-line complexity
+  app.use((req, res, next) => {
+    try {
+      const methodAndPathKey = getMethod(req);
+      if (methodAndPathKey && methodAndPathKey.method && methodAndPathKey.pathKey) {
+        const method = methodAndPathKey.method;
+        updateSchemesAndHost(req);
+        processors.processPath(req, method, methodAndPathKey.pathKey);
+        processors.processHeaders(req, method, spec);
+        processors.processBody(req, method);
+        processors.processQuery(req, method);
+      }
+    } catch (e) {
+      /** TODO - shouldn't we do something here? */
+    } finally {
       next();
-    });
-    init(aApiDocsPath);
+    }
+  });
+}
+
+/**
+ * TODO
+ *
+ * 1. Rename the parameter names
+ * (this will require better global variable naming to allow re-assigning properly)
+ *
+ * 2. the `aPath` is ignored only if it's `undefined`,
+ * but if it's set to an empty string `''`, then the tests fail.
+ * I think we should be checking for falsy values, not only `undefined` ones.
+ *
+ * 3. (Breaking) Use object for optional parameters:
+ * https://github.com/mpashkovskiy/express-oas-generator/issues/35
+ *
+ */
+/**
+ * @warn it's preferred that you use `injectResponseMiddleware`,
+ * `injectRequestMiddleware` and `serveApiDocs` **individually**
+ * and not directly from this `init` function,
+ * because we need `injectRequestMiddleware` to be placed as the
+ * very last middleware and we cannot guarantee this here,
+ * since we're only using an arbitrary setTimeout of `1000` ms.
+ *
+ * See
+ * https://github.com/mpashkovskiy/express-oas-generator/pull/32#issuecomment-546807216
+ *
+ * @description initialize the `express-oas-generator`.
+ *
+ * This will apply both `injectResponseMiddleware` and `injectRequestMiddleware`
+ * and also will call `serveApiDocs`.
+ *
+ * @param {Express} aApp - the express app
+ * @param {*} [aPredefinedSpec=undefined]
+ * @param {string|undefined} [aPath=undefined] where to write the openAPI specification to.
+ * Specify this to create the openAPI specification file.
+ * @param {number} [aWriteInterval=10000] how often to write the openAPI specification to file
+ * @param {string} [aApiDocsPath=api-docs] where to serve the openAPI docs. Defaults to `api-docs`
+ */
+function init(aApp, aPredefinedSpec = undefined, aPath = undefined, aWriteInterval = 1000 * 10, aApiDocsPath = 'api-docs') {
+  /**
+   * TODO - shouldn't `predefinedSpec` be assigned @ `serveApiDocs`?
+   *
+   * I don't know if the `predefinedSpec` is used anywhere for `injectResponseMiddleware`
+   * and before `injectRequestMiddleware` is called
+   */
+  predefinedSpec = aPredefinedSpec;
```

Comment:
I'd like to solve this - I think I probably should place the `predefinedSpec` assignment to the global variable into the `serApiDocs` function.
I'm passing it as a parameter currently.
If it's safe and no-one else needs it, then we should do that for clarity.

Reply:
Allright, this is a bit complex.

First of the all we need an extra test for case where we apply middlewares individually. We need it it to prove that this approach works. Just add a test case like:

```javascript
  const path = 'hello';
  const value = 2;
  const app = express();
  generator.injectResponseMiddleware(app);
  app.get(path, (req, res, next) => {
    res.setHeader('Content-Type', 'text/plain');
    res.send(PLAIN_TEXT_RESPONSE);
    return next();
  });
  app.set('port', port);
  const server = app.listen(app.get('port'), function() {
      generator.injectRequestMiddleware();
      request.get(`http://localhost:${port}/${path}?a=${value}`, () => {
        const spec = generator.getSpec();
        expect(spec.paths[path].get.parameters[0].example).toBe(value);
        server.close();
        done();
      });
  });
```

And that will fail badly because `serveApiDocs` won't be called here at all and no spec initialization won't happen!

So:
* at least one extra test case is needed to test new approach
* `serveApiDocs` should be placed in `injectRequestMiddleware()`
* global variable `predefinedSpec` is needed because it is used for on-the-fly generation of the spec in middlewares during requests see lines 95, 99
* assignment to global variable `predefinedSpec` could be done in `injectRequestMiddleware` or in `serveApiDocs`

Reply:
Ah, I see that the name I chose (`serveApiDocs`) is misleading -- it also plays with the global `spec` variable.

1. Should I refactor `serveApiDocs` into 2 separate functions - one for initializing the spec, and the other for actually serving the api docs?

I think I should keep only this part as the `serveApiDocs`:

https://github.com/mpashkovskiy/express-oas-generator/pull/39/files/11872fad23ff084989f425d5e578a65efda8240d#diff-168726dbe96b3ce427e7fedce31bb0bcR93-R100

And then move all of this into, say `createInitialSpec` or `initializeSpec`:

https://github.com/mpashkovskiy/express-oas-generator/pull/39/files/11872fad23ff084989f425d5e578a65efda8240d#diff-168726dbe96b3ce427e7fedce31bb0bcR56-R92

---

2. Also, instead of using `patchSpec(predefinedSpec)` in [here](https://github.com/mpashkovskiy/express-oas-generator/pull/39/files/11872fad23ff084989f425d5e578a65efda8240d#diff-168726dbe96b3ce427e7fedce31bb0bcR95) and [here](https://github.com/mpashkovskiy/express-oas-generator/pull/39/files/11872fad23ff084989f425d5e578a65efda8240d#diff-168726dbe96b3ce427e7fedce31bb0bcR99), couldn't we just use the global `spec` variable, since we've already patched it? Or is there a difference?

---

The rest is clear, I just don't fully understand the test case yet, I'll see:D